odl/phantom/geometric.py

```diff
@@ -703,6 +704,232 @@ def ellipsoid_phantom(space, ellipsoids, min_pt=None, max_pt=None):
+        Uniformly discretized space in which the phantom should be generated.
+        If ``space.shape`` is 1 in an axis, a corresponding slice of the
+        phantom is created (instead of squashing the whole phantom into the
+        slice).
+    cuboids : list of lists
+        Each row should contain the entries ::
+
+            'value',
+            'axis_1', 'axis_2',
+            'center_x', 'center_y',
+            'rotation'
+
+        The provided cuboids need to be specified relative to the
+        reference rectangle ``[-1, -1] x [1, 1]``. Angles are to be given
+        in radians.
+
+    Returns
+    -------
+    phantom : ``space`` element
+        2D cuboids phantom in ``space``.
+    """
+    # Blank image
+    p = np.zeros(space.shape, dtype=space.dtype)
+
+    minp = space.grid.min_pt
+    maxp = space.grid.max_pt
+
+    # Create the pixel grid
+    grid_in = space.grid.meshgrid
+
+    # move points to [-1, 1]
+    grid = []
+    for i in range(2):
+        mean_i = (minp[i] + maxp[i]) / 2.0
+        # Where space.shape = 1, we have minp = maxp, so we set diff_i = 1
+        # to avoid division by zero. Effectively, this allows constructing
+        # a slice of a 2D phantom.
+        diff_i = (maxp[i] - minp[i]) / 2.0 or 1.0
+        grid.append((grid_in[i] - mean_i) / diff_i)
+
+    for cuboid in cuboids:
+        assert len(cuboid) == 6
+
+        intensity = cuboid[0]
+        a_squared = cuboid[1] ** 2
+        b_squared = cuboid[2] ** 2
+        x0 = cuboid[3]
+        y0 = cuboid[4]
+        theta = cuboid[5]
+
+        scales = [1 / a_squared, 1 / b_squared]
+        center = (np.array([x0, y0]) + 1.0) / 2.0
+
+        # Create the offset x,y and z values for the grid
+        if theta != 0:
+            # Rotate the points to the expected coordinate system.
+            ctheta = np.cos(theta)
+            stheta = np.sin(theta)
+
+            mat = np.array([[ctheta, stheta],
+                            [-stheta, ctheta]])
+
+            # Calculate the points that could possibly be inside the volume
+            # Since the points are rotated, we cannot do anything directional
+            # without more logic
+            max_radius = np.abs(mat).dot(np.sqrt([a_squared, b_squared]))
+            idx, shapes = _getshapes_2d(center, max_radius, space.shape)
+
+            subgrid = [g[idi] for g, idi in zip(grid, shapes)]
+            offset_points = [vec * (xi - x0i)[..., None]
+                             for xi, vec, x0i in zip(subgrid, mat.T, [x0, y0])]
+            rotated = offset_points[0] + offset_points[1]
+            np.square(rotated, out=rotated)
+            max_distance = np.maximum(scales[0] * rotated[..., 0],
+                                      scales[1] * rotated[..., 1])
+        else:
+            # Calculate the points that could possibly be inside the volume
+            max_radius = np.sqrt([a_squared, b_squared])
+            idx, shapes = _getshapes_2d(center, max_radius, space.shape)
```

Comment:
Why do you have to do this for non-rotated rectangles? Can't you just get the min and max per axis and use those directly as bounding boxes? In other words, can't you directly construct `slice` objects from the min and max and use those for indexing?

Reply:
OTOH, I can see why you want to use the same kind of logic. Maybe it's okay.

Reply:
I'll just leave this as is since the other optimization is relatively minor for most cases. This would only give a notable improvement in the case of slightly rotated and very elongated rectangles, which I'll just leave for now.